presto-geospatial/src/main/java/com/facebook/presto/geospatial/BingTileFunctions.java

```diff
@@ -174,20 +177,129 @@ public static Block geometryToBingTiles(@SqlType(GEOMETRY_TYPE_NAME) Slice input
         // XY coordinates start at (0,0) in the left upper corner and increase left to right and top to bottom
         int tileCount = toIntExact((rightLowerTile.getX() - leftUpperTile.getX() + 1) * (rightLowerTile.getY() - leftUpperTile.getY() + 1));
 
-        BlockBuilder blockBuilder = BIGINT.createBlockBuilder(new BlockBuilderStatus(), tileCount);
-        for (int x = leftUpperTile.getX(); x <= rightLowerTile.getX(); x++) {
-            for (int y = leftUpperTile.getY(); y <= rightLowerTile.getY(); y++) {
-                BingTile tile = BingTile.fromCoordinates(x, y, zoomLevel);
+        if (pointOrRectangle) {
+            checkCondition(tileCount <= 1_000_000, "The number of input tiles is too large (more than 1M) to compute a set of covering bing tiles.");
+        }
+        else {
+            long complexity = tileCount * getPointCount(geometry.getEsriGeometry());
+            checkCondition(complexity <= 25_000_000, "The zoom level is too high or the geometry is too complex to compute a set of covering bing tiles. " +
+                    "Please use a lower zoom level or convert the geometry to its bounding box using the ST_Envelope function.");
+        }
 
-                if (pointOrRectangle || !tileToPolygon(tile).disjoint(geometry)) {
-                    BIGINT.writeLong(blockBuilder, tile.encode());
+        BlockBuilder blockBuilder = BIGINT.createBlockBuilder(null, tileCount);
```

Comment:
The `geometryToBingTiles` method is now a bit on the long side, how about extracting a new method that does the actual tiling (covering the if/else block below) and here at L189 we just return a call to that method (e.g., `return geometryToBingTiles(...);`? That will simplify this method a bit.